### ISSUE TITLE
Priorities for 2016

### ISSUE THREAD

**Issue by ghedipunk** [ghedipunk/PHP-Websockets#67]
First priority: Finish TLS support in the Legacy branch, merge to Master.

High priorities: 
 * Unit testing in new development branch.
 * Autobahn testsuite.
 * Minimum Viable Product for new development branch.
 * Project site.  (Download zip/tarball, project announcements, documentation wiki, reference implementations (chat, simple games, CMS features), support forum... Start simple, of course.)

Normal priorities:
 * Parse cookies from headers
 * Expose PHP session
 * LibEv
 * Re-optimize everything

Potential features:
 * Composer support

Please add any priorities that I missed, or feature requests, in the comments below.  This first post will be regularly updated.

**Comment by Xaraknid** [comment]
Normal priorities:
As you plan to add libev you can reintroduce socket as they work on same level.
Also adding libevent will be great for windows server use this library http://docs.php.net/event most recently active and with more chance to see php 7.

**Comment by ghedipunk** [comment]
Changes made from original version:

Added 
 * Add support for multiple apps
 * Add support for different network IO models (i.e., be able to use both Berkeley and Stream sockets interchangeably, different TLS settings).
 * Add support for binding to multiple TCP/IP sockets (Multi-NIC, IPv4 + IPv6 at the same time, multiple port numbers, etc., see below)
 * Add the ability to choose different event loops, such as libevent (http://docs.php.net/event) or a while loop around socket_select(), as well as allow user-defined loops.

Added section for recurring priorities.  One big difference between this WS library and other WS libraries is how easy it is to get working... Especially that it works as a basic echo server out of the box.  Ease of implementation is a strength that we need to maintain, and reviewing that ease of implementation each time we consider a new feature will help keep us on track.

Now, to that "see below" part of binding to multiple sockets, an overly complex example:

*Scenario:*

Server sitting at a DMZ with two Network Interface Cards (NICs). One NIC is accessible through a public network, has the IP of `12.34.56.78`.  The other NIC is on the private network with an IP of `192.168.0.100`.  

The server has a WebSockets server running that allows users to chat while they're in the office or using a VPN connection.  There are also daemons on that server that monitor other message queues, such as whether there are blog comments flagged as spam in the company's blog, and the WS server is used to relay these notifications to subscribed users.

Thus, the application has the requirements that it be accessible through both localhost and through its internal network NIC, but must not listen to the public network NIC.

(I know what you're thinking...  You're about to say "ah, but that's what firewalls and router ACLs are for!"  Yes, you're right, the firewalls and routers should all be configured to block this traffic, but this is _also_ what listen addresses are for.  If you've ever watched the first Shrek movie, remember, security is like an ogre: they both have layers.  Just like onions.  The more layers of security you have, the less likely it is that you'll be leaking sensitive internal communications when your firewall hits a failure state and fails open, or your sysadmin slips up and writes a logical error into the ACL that effectively bypasses all DENY steps.  Yes, cakes have layers, too, but security isn't like cake.  People like cake.  People don't like security or ogres.  Or onions. /rant> )

So the solution? Have the server bind to 127.0.0.1:8080 and 192.168.0.100:8080 explicitly, and leave out all other address and port combinations.  (And yes, _please_ configure your firewalls and ACLs, and _please_ get a second pair of eyes on those settings, and _please_ test that people external to your network don't even see that the port might be in use anywhere.)

That's my primary justification for wanting multiple distinct TCP/IP listen sockets.  There's also a secondary justification, in that we can then split up WS and WSS traffic, mix IPv4 and IPv6 traffic as well, and could even have different apps listen to different message queues based on things like address and port.

**Comment by Xaraknid** [comment]
I think it's bad idea to have ws and wss executed by the same process unlike http protocol, ws connection stay alive so mixing unsecure connection with secure is a leak of security. Those 2 should and must be run by different process.

By openning 0.0.0.0:8080 I can connect to my server from 2 different internet adress ip ( having each their domainname) and localhost .  With information from handshake I can dispatch the message to everyone or only a portion connected from ( example.com), can go as deep as /ressource.

example.com:8080/echobot
example2.com:8080/echobot
localhost:8080/echobot

The challenge will be to see IPv4 mix with IPv6.

**Comment by ghedipunk** [comment]
We could have it so that the user would have to pick only one network IO model, but still have multiple choices, or specify their own.  Binding to sockets and actually dealing with the traffic sent across them are different enough that we can have the best of both worlds.

Thus, they can have both IPv4 and IPv6 addresses, or specify a list of addresses in case they want to exclude a NIC as in the case of a server sitting in a DMZ and want/need to avoid listening on 0.0.0.0.  But if they wanted to have a TLS and a separate non-TLS connection, they would need different processes, since they server makes you choose which network IO model to use (or take the default) and won't let you change once it starts listening for connections.

**Comment by ghedipunk** [comment]
Added under ideas to consider:

Package management system.

Obviously far more involved than just maintaining a server library.  Would need a baseline repository and a way for people to define their own custom repo, as well as a script to handle managing the packages on the user's server.

**Comment by Xaraknid** [comment]
What i mean by process is you must run 2 independant server php unsecure.php and php secure.php.  This prevent data from secure client connection to be sent intentionnaly or not to a unsecure client vice-versa.  Server side data is decrypted and is process plain text, so it's easy to send data from secure to unsecure client.

**Comment by ghedipunk** [comment]
Yes, that matches my definition of process as well, and you made a great point about not mixing TLS and non-TLS connections in the same process.  I've changed the top post with the requirement that once we start listening for traffic, we can't switch to another way to accept traffic, so if we start with TLS, we have to stay on TLS.

**Comment by Xaraknid** [comment]
For my personnal need I retraint acess from apps to core method directly.  I switch alot of method from protected to private ( all method that are use internally ).

Apps work with a jailed list of client, in other word client from /echobot didn't know of any client from /news vice-versa.  Only core have the complete list of clients and because I jailed core apps didn't know either what inside core but core know everything but only act as dispatcher.

**Comment by Xaraknid** [comment]
After some tought it's easy task to listen on multiple ip/port with stream and socket at least the challenge will be libev. I will do some test and come back with that.

**Comment by Xaraknid** [comment]
High priorities : use NON BLOCKING instead of BLOCKING connection.  Will solve many problems.
